File: webflux-dialogue/src/test/java/com/study/webflux/rag/application/dialogue/pipeline/stage/SystemPromptServiceTest.java
package com.study.webflux.rag.application.dialogue.pipeline.stage;

import java.util.List;

import com.study.webflux.rag.application.dialogue.policy.PromptTemplatePolicy;
import com.study.webflux.rag.domain.dialogue.model.ConversationSessionId;
import com.study.webflux.rag.domain.dialogue.model.PersonaId;
import com.study.webflux.rag.domain.dialogue.port.TemplateLoaderPort;
import com.study.webflux.rag.domain.memory.model.Memory;
import com.study.webflux.rag.domain.memory.model.MemoryRetrievalResult;
import com.study.webflux.rag.domain.memory.model.MemoryType;
import com.study.webflux.rag.domain.retrieval.model.RetrievalContext;
import com.study.webflux.rag.domain.retrieval.model.RetrievalDocument;
import com.study.webflux.rag.fixture.ConversationSessionFixture;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class SystemPromptServiceTest {

	@Mock
	private TemplateLoaderPort templateLoader;

	@Test
	@DisplayName("시스템 프롬프트 생성 시 모든 섹션을 포함한다")
	void buildSystemPrompt_shouldRenderBaseTemplateWithAllSections() {
		when(templateLoader.load("system/persona/maid")).thenReturn("persona");

		PromptTemplatePolicy policy = new PromptTemplatePolicy(
			"{{persona}}\n\n{{common}}\n\n{{memories}}\n\n{{context}}",
			"",
			"common",
			null);

		SystemPromptService service = new SystemPromptService(templateLoader, policy);

		ConversationSessionId sessionId = ConversationSessionFixture.createId();
		Memory experiential = Memory.create(sessionId, MemoryType.EXPERIENTIAL, "사용자는 러닝을 좋아한다.",
			0.8f);
		Memory factual = Memory.create(sessionId, MemoryType.FACTUAL, "사용자의 직업은 개발자다.", 0.9f);
		MemoryRetrievalResult memories = MemoryRetrievalResult.of(List.of(experiential),
			List.of(factual));
		RetrievalContext context = RetrievalContext.of("사용자 취미",
			List.of(RetrievalDocument.of("러닝은 체력 향상에 도움이 된다.", 90)));

		String prompt = service.buildSystemPrompt(PersonaId.of("maid"), context, memories);

		assertThat(prompt).contains("persona").contains("common").contains("대화 상대에 대한 기억:")
			.contains("경험적 기억:").contains("사실 기반 기억:")
			.contains("- 사용자는 러닝을 좋아한다.").contains("- 사용자의 직업은 개발자다.")
			.contains("참고 정보:").contains("러닝은 체력 향상에 도움이 된다.");
	}

	@Test
	@DisplayName("페르소나 템플릿 누락 시 기본 프롬프트로 폴백한다")
	void buildSystemPrompt_shouldFallbackToConfiguredSystemPromptWhenPersonaTemplateIsMissing() {
		when(templateLoader.load("system/persona/maid"))
			.thenThrow(new RuntimeException("persona missing"));

		PromptTemplatePolicy policy = new PromptTemplatePolicy(
			"",
			"",
			"common",
			"configured persona");

		SystemPromptService service = new SystemPromptService(templateLoader, policy);

		String prompt = service.buildSystemPrompt(PersonaId.of("maid"),
			RetrievalContext.empty("query"),
			MemoryRetrievalResult.empty());

		assertThat(prompt).isEqualTo("configured persona\n\ncommon");
	}
}

File: app-miyou/src/test/kotlin/com/miyou/app/application/dialogue/pipeline/stage/SystemPromptServiceTest.kt
package com.miyou.app.application.dialogue.pipeline.stage

import com.miyou.app.application.dialogue.policy.PromptTemplatePolicy
import com.miyou.app.domain.dialogue.model.PersonaId
import com.miyou.app.domain.dialogue.port.TemplateLoaderPort
import com.miyou.app.domain.memory.model.Memory
import com.miyou.app.domain.memory.model.MemoryRetrievalResult
import com.miyou.app.domain.memory.model.MemoryType
import com.miyou.app.domain.retrieval.model.RetrievalContext
import com.miyou.app.domain.retrieval.model.RetrievalDocument
import com.miyou.app.fixture.ConversationSessionFixture
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.junit.jupiter.MockitoExtension

@ExtendWith(MockitoExtension::class)
class SystemPromptServiceTest {
    @Mock
    private lateinit var templateLoader: TemplateLoaderPort

    @Test
    @DisplayName("시스템 프롬프트를 구성된 모든 섹션으로 렌더링한다")
    fun buildSystemPrompt_shouldRenderBaseTemplateWithAllSections() {
        `when`(templateLoader.load("system/persona/maid")).thenReturn("persona")

        val policy =
            PromptTemplatePolicy(
                "{{persona}}\n\n{{common}}\n\n{{memories}}\n\n{{context}}",
                "",
                "common",
                "",
            )

        val service = SystemPromptService(templateLoader, policy)
        val sessionId = ConversationSessionFixture.createId()
        val experiential = Memory.create(sessionId, MemoryType.EXPERIENTIAL, "user likes sushi", 0.8f)
        val factual = Memory.create(sessionId, MemoryType.FACTUAL, "user is a developer", 0.9f)
        val memories = MemoryRetrievalResult.of(listOf(experiential), listOf(factual))
        val context =
            RetrievalContext.of(
                "user preference",
                listOf(RetrievalDocument.of("sushi improves stamina", 90)),
            )

        val prompt = service.buildSystemPrompt(PersonaId.of("maid"), context, memories)

        assertThat(prompt)
            .contains("persona")
            .contains("common")
            .contains("- user likes sushi")
            .contains("- user is a developer")
            .contains("sushi improves stamina")
    }

    @Test
    @DisplayName("페르소나 템플릿이 없으면 설정된 기본 프롬프트로 대체한다")
    fun buildSystemPrompt_shouldFallbackToConfiguredSystemPromptWhenPersonaTemplateIsMissing() {
        `when`(templateLoader.load("system/persona/maid"))
            .thenThrow(RuntimeException("persona missing"))

        val policy = PromptTemplatePolicy("", "", "common", "configured persona")
        val service = SystemPromptService(templateLoader, policy)

        val prompt =
            service.buildSystemPrompt(
                PersonaId.of("maid"),
                RetrievalContext.empty("query"),
                MemoryRetrievalResult.empty(),
            )

        assertThat(prompt).isEqualTo("configured persona\n\ncommon")
    }

    @Test
    @DisplayName("정책 템플릿이 비어 있어도 시스템 프롬프트를 생성한다")
    fun buildSystemPrompt_shouldHandleBlankPolicyTemplates() {
        val policy = PromptTemplatePolicy("", "", "", " configured persona ")
        val service = SystemPromptService(templateLoader, policy)

        val prompt =
            service.buildSystemPrompt(
                PersonaId.defaultPersona(),
                RetrievalContext.empty("query"),
                MemoryRetrievalResult.empty(),
            )

        assertThat(prompt).isEqualTo("configured persona")
    }
}
